### PR TITLE
[SPARK-11425] [SPARK-11486] Improve hybrid aggregation

### PR DESCRIPTION
**Pull request by davies**
After aggregation, the dataset could be smaller than inputs, so it's better to do hash based aggregation for all inputs, then using sort based aggregation to merge them.

### CODE REVIEW

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 1]
```diff
@@ -860,45 +844,48 @@ void growAndRehash() {
...
 
     if (enablePerfMetrics) {
       timeSpentResizingNs += System.nanoTime() - resizeStartTime;
     }
   }
+
+  /**
+   * Return the compacted long array.
+   */
+  public long[] getCompactArray() {
```
Comment (viirya) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: After call `getCompactArray`, the content of `longArray` is modified. Can this `BytesToBytesMap` be normally used later? Because the position in `longArray` for a key should be determined by `(keyBase, keyOffset, keyLength)`. If the positions are modified, can the methods such as `safeLookup` work?

Comment (davies) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: No, after this, the map is broken, should be freed later.

Comment (viirya) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Is it needed to add comment for it?

**Comment by viirya** [comment]
Besides, as we discussed in #9067, should we add a configuration for turning on/off this feature? This feature may not always have performance gain.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 2]
```diff
@@ -860,45 +844,48 @@ void growAndRehash() {
...
-    // Allocate the new data structures
-    try {
-      allocate(Math.min(growthStrategy.nextCapacity(oldCapacity), MAX_CAPACITY));
-    } catch (OutOfMemoryError oom) {
-      longArray = oldLongArray;
-      bitset = oldBitSet;
-      throw oom;
-    }
+    final long[] oldLongArray = longArray;
+    allocate(Math.min(growthStrategy.nextCapacity(longArray.length / 2), MAX_CAPACITY));
```
Comment (viirya) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: Don't we need to catch `OutOfMemoryError` anymore?

**Comment by davies** [comment]
Currently, the old one is broken, I'd to remove that one. The new should be as fast as the old one in worst case, I think we don't need a configuration for this.

**Comment by davies** [comment]
After some benchmark, realized that using hashcode as prefix in timsort will cause regression in timsort and snappy compression (especially for aggregation after join, the order of records will become random). I will revert that part.

benchmark code:
```
sqlContext.setConf("spark.sql.shuffle.partitions", "1")
N = 1<<25
M = 1<<20
df = sqlContext.range(N).selectExpr("id", "repeat(id, 2) as s")
df.show()
df2 = df.select(df.id.alias('id2'), df.s.alias('s2'))
j = df.join(df2, df.id==df2.id2).groupBy(df.s).max("id", "id2")
n = j.count()
```

Another interesting finding is that Snappy will slowdown the spilling by 50% of end-to-end time, LZ4 will be faster than Snappy, but still 10% slower than not-compressed. Should we use `false` as the default value for `spark.shuffle.spill.compress`?(PS: tested on Mac with SSD, it may not be true on HD)

**Comment by davies** [comment]
ping @yhuai @JoshRosen

**Comment by JoshRosen** [comment]
> Currently, the old one is broken, I'd to remove that one.

@davies, are you referring to the old Aggregate1 interface or the old implementation of sort fallback here?

**Comment by JoshRosen** [comment]
@davies, the block comment at the top of `TungstenAggregationIterator` is now out-of-date; do you mind updating it to reflect the new behavior?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 3]
```diff
@@ -386,6 +386,18 @@ public void insertKVRecord(Object keyBase, long keyOffset, int keyLen,
...
   /**
+   * Merges another UnsafeExternalSorters into this one, the other one will be emptied.
+   *
+   * @throws IOException
+   */
+  public void merge(UnsafeExternalSorter other) throws IOException {
+    other.spill();
+    spillWriters.addAll(other.spillWriters);
+    other.spillWriters.clear();
+    other.cleanupResources();
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: Ordinarily, this would end up deleting the spill files, but it doesn't because of the `spillWriters.clear()` call above. If you end up updating this patch, mind adding a one-line comment to explain this (since it's a subtle point)?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 4]
```diff
@@ -762,15 +679,7 @@ class TungstenAggregationIterator(
...
-  testFallbackStartsAt match {
-    case None =>
-      processInputs()
-    case Some(fallbackStartsAt) =>
-      // This is the testing path. processInputsWithControlledFallback is same as processInputs
-      // except that it switches to sort-based aggregation after `fallbackStartsAt` input rows
-      // have been processed.
-      processInputsWithControlledFallback(fallbackStartsAt)
-  }
+  processInputs(testFallbackStartsAt.getOrElse(Int.MaxValue))
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I guess it isn't a huge deal to have a really high limit here, as opposed to no limit, because the cost of spilling after processing a couple of billion rows will be fairly minimal; either we would have spilled earlier or we're experiencing a huge reduction factor, so the cost of spilling should be minimal.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 5]
```diff
@@ -473,24 +473,28 @@ class TungstenAggregationIterator(
...
-    1024 * 16, // initial capacity
-    TaskContext.get().taskMemoryManager().pageSizeBytes,
-    false // disable tracking of performance metrics
-  )
+  private[this] var hashMap = createHashMap()
 
   // The function used to read and process input rows. When processing input rows,
   // it first uses hash-based aggregation by putting groups and their buffers in
   // hashMap. If we could not allocate more memory for the map, we switch to
   // sort-based aggregation (by calling switchToSortBasedAggregation).
```
Comment (JoshRosen) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Update comment to reflect the fact that we use multiple hash-maps, spilling after each becomes full then using sort to merge the spills?

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java [hunk 6]
```diff
@@ -142,6 +142,10 @@ public void insertKV(UnsafeRow key, UnsafeRow value) throws IOException {
       value.getBaseObject(), value.getBaseOffset(), value.getSizeInBytes(), prefix);
   }
 
+  public void merge(UnsafeKVExternalSorter other) throws IOException {
```
Comment (JoshRosen) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: Add comment?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala [hunk 7]
```diff
@@ -291,28 +278,27 @@ class UnsafeFixedWidthAggregationMapSuite
...
       }
     }
+    val sorter2 = map.destructAndCreateExternalSorter()
+    sorter.merge(sorter2)
 
     var count = 0
     val iter = sorter.sortedIterator()
     while (iter.next()) {
-      // At here, we also test if copy is correct.
-      iter.getKey.copy()
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: Why remove this?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala [hunk 8]
```diff
@@ -291,28 +278,27 @@ class UnsafeFixedWidthAggregationMapSuite
...
-      iter.getValue.copy()
       count += 1
     }
 
-    // 1 record was from the map and 4096 records were explicitly inserted.
-    assert(count === 4097)
-
-    map.free()
+    // 1 record in one map.
+    assert(count === 42)
```
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: Why this change? Shouldn't the total number of records have been unchanged by this refactoring?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 4 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: The reason I'd like to have a really high limit is to avoid the branch, like `testFallbackStartsAt.isDefined() && i < testFallbackStartsAt.get()`. It's a common trick, I think.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala (hunk 8 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: The sorter will not combine the multiple records with same keys, but map will.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala (hunk 7 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: Could add that back.

**Comment by davies** [comment]
@JoshRosen I meant the old implementation of switching, it's broken in master (not 1.5 branch), had not figured out the root cause.

**Comment by JoshRosen** [comment]
Do you remember what the symptoms were? Just curious.

**Comment by davies** [comment]
@JoshRosen When the grouping key is StringType, but no column in buffer, then got:
```
java.lang.UnsupportedOperationException
	at org.apache.spark.sql.catalyst.expressions.UnsafeRow.update(UnsafeRow.java:193)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificMutableProjection.apply(generated.java:40)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregationIterator.switchToSortBasedAggregation(TungstenAggregationIterator.scala:643)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregationIterator.processInputs(TungstenAggregationIterator.scala:517)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregationIterator.start(TungstenAggregationIterator.scala:779)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregate$$anonfun$doExecute$1.org$apache$spark$sql$execution$aggregate$TungstenAggregate$$anonfun$$executePartition$1(TungstenAggregate.scala:128)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregate$$anonfun$doExecute$1$$anonfun$3.apply(TungstenAggregate.scala:137)
	at org.apache.spark.sql.execution.aggregate.TungstenAggregate$$anonfun$doExecute$1$$anonfun$3.apply(TungstenAggregate.scala:137)
	at org.apache.spark.rdd.MapPartitionsWithPreparationRDD.compute(MapPartitionsWithPreparationRDD.scala:64)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:300)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:264)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:300)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:264)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:73)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
	at org.apache.spark.scheduler.Task.run(Task.scala:88)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:214)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
15/10/28 23:25:08 DEBUG TaskSchedulerImpl: parentName: , name: TaskSet_3, runningTasks: 0
```

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala (hunk 8 above)
Comment (JoshRosen) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: Ah. Update comment to clarify that we spilled 42 times, hence this number?

**Comment by yhuai** [comment]
@JoshRosen @davies has a case to reproduce that problem

**Comment by JoshRosen** [comment]
@davies: in the stacktrace that you posted, it looks like it's failing when trying to update a string column when extracting a partial aggregate result from an input row. What's confusing to me, though, is the fact that a string column in a partial aggregate result buffer should have prevented TungstenAggregate from being planned in the first place (see `TungstenAggregate.supportsAggregate()`). Do you think that there might be an input attribute binding error that's causing this (i.e. we mistakenly treat grouping columns as belonging to the partial aggregate result)?

**Comment by JoshRosen** [comment]
I _do_ think that the change in this patch makes sense, but I'd like to really get to the bottom of why the old code malfunctioned.

**Comment by davies** [comment]
This failure could be reproduce by 
```
sqlContext.range(N).selectExpr("id", "repeat(id, 20) as s").groupBy("s").max("id").count()
```

Maybe related to this
https://github.com/apache/spark/pull/9038/files#diff-b83a65c9afcef83e053ecc0afecba1a4R633

**Comment by yhuai** [comment]
@JoshRosen The cause is that if the buffer is an empty one (e.g. we do not have agg function or all agg functions do not use any buffer slot), the initialInputBufferOffset is still 0 and when the key is a string, the following code will trigger the exception. Unfortunately, our test only use int columns as the key columns.

```
newMutableProjection(
  originalInputAttributes.drop(initialInputBufferOffset),
  originalInputAttributes)()
```

**Comment by yhuai** [comment]
Maybe `sqlContext.range(100000).selectExpr("id", "repeat(id, 20) as s").groupBy("s").agg($"s")` is a better case.

**Comment by JoshRosen** [comment]
What do you think about adding a dedicated regression test for that one case as part of a separate PR? I'll try that now.

**Comment by davies** [comment]
Will add that into this PR, working on it.

**Comment by JoshRosen** [comment]
I filed https://issues.apache.org/jira/browse/SPARK-11486 to describe the String aggregation bug.

**Comment by JoshRosen** [comment]
@davies, @yhuai: I tried taking two sample regression test queries that you listed here and putting them into AggregationQuerySuite, which should automatically trigger spilling after a small number of records, but I wasn't able to reproduce the exception here. How did you get this to fail?

I want to make sure that the regression test that we add is actually capable of failing prior to this change.

**Comment by yhuai** [comment]
So, you tried `TungstenAggregationQueryWithControlledFallbackSuite` and it did not fail?

**Comment by davies** [comment]
I tried with master, it did failed with the sam exception:
```

[info] *** 1 TEST FAILED ***
[error] Failed: Total 6, Failed 1, Errors 0, Passed 5
[error] Failed tests:
[error] 	org.apache.spark.sql.hive.execution.TungstenAggregationQueryWithControlledFallbackSuite
```

**Comment by JoshRosen** [comment]
There was a problem in my test-selection regex... trying again.

**Comment by JoshRosen** [comment]
Yep, confirmed and reproduced. If we just wanted to address that bug, we could do something like #9447, but I think the broader set of changes here are a good idea.

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 4 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: It still possible that we basically have lot of memory space to use, right?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Each record needs 30+ bytes, it needs to have more than 60G memory for single task to trigger this spilling, I think that's fine.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala [hunk 9]
```diff
@@ -502,44 +511,37 @@ class TungstenAggregationIterator(
...
-        } else {
-          processRow(buffer, newInput)
+          val sorter = hashMap.destructAndCreateExternalSorter()
+          if (externalSorter == null) {
+            externalSorter = sorter
+          } else {
+            externalSorter.merge(sorter)
+          }
+          i = 0
+          hashMap = createHashMap()
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: Before we create the new map, I guess it will not hurt if we call `hashMap`? We just make our clear that the previous map will be freed.

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: I mean from this part of code, it is not obvious when we free the map.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala [hunk 10]
```diff
@@ -183,31 +180,30 @@ class UnsafeFixedWidthAggregationMapSuite
...
       if ((i % 100) == 0) {
-        memoryManager.markExecutionAsOutOfMemoryOnce()
-        sorter.closeCurrentPage()
+        val sorter2 = map.destructAndCreateExternalSorter()
+        sorter.merge(sorter2)
+        map = createMap()
       }
     }
+    val sorter2 = map.destructAndCreateExternalSorter()
+    sorter.merge(sorter2)
```
Comment (yhuai) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: Looks like this updated test is not doing the exactly same thing with the previous version? In the old version, after we get the sorter we add new records to the sorter directly. In this updated version, we will always use merge to add spilled files to the sorter. Should we just create new tests for this new behavior (using `merge`)?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala [hunk 11]
```diff
@@ -589,6 +588,13 @@ abstract class AggregationQuerySuite extends QueryTest with SQLTestUtils with Te
     }
   }
 
+  test("no aggregation function") {
+    val df = sqlContext.range(20).selectExpr("id", "repeat(id, 1) as s")
+      .groupBy("s").count()
+      .groupBy().count()
+    checkAnswer(df, Row(20) :: Nil)
+  }
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: Maybe add a description of the original problem at here or just add a link?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala (hunk 9 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala]: It's freed in destructAndCreateExternalSorter, I think that's the mean of `destruct`.

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java [hunk 12]
```diff
@@ -386,24 +388,37 @@ public void insertKVRecord(Object keyBase, long keyOffset, int keyLen,
...
+      readingIterator = new SpillableIterator(inMemSorter.getSortedIterator());
       return readingIterator;
     } else {
       final UnsafeSorterSpillMerger spillMerger =
-        new UnsafeSorterSpillMerger(recordComparator, prefixComparator, numIteratorsToMerge);
+        new UnsafeSorterSpillMerger(recordComparator, prefixComparator, spillWriters.size());
       for (UnsafeSorterSpillWriter spillWriter : spillWriters) {
         spillMerger.addSpillIfNotEmpty(spillWriter.getReader(blockManager));
       }
-      spillWriters.clear();
```
Comment (davies) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: @JoshRosen Do you remember why we need to clear this? Once cleared, how to delete the spilled files?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala (hunk 11 above)
Comment (davies) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/AggregationQuerySuite.scala]: Added JIRA number

**Comment by davies** [comment]
@JoshRosen @yhuai pushed a refactor on this (reduce possibility of full GC by re-use the array in map), please take another look.

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 13]
```diff
@@ -647,7 +648,7 @@ public boolean putNewKey(Object keyBase, long keyOffset, int keyLength,
       assert(bitset != null);
       assert(longArray != null);
 
-      if (numElements == MAX_CAPACITY || !canGrowArray) {
+      if (numElements == MAX_CAPACITY || numElements > growthThreshold && !canGrowArray) {
```
Comment (yhuai) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: When we have `numElements  <= growthThreshold && !canGrowArray`, it is guaranteed that our page still has space to put this key, right?

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java [hunk 14]
```diff
@@ -83,11 +83,8 @@ public UnsafeKVExternalSorter(
         /* initialSize */ 4096,
         pageSizeBytes);
     } else {
-      // The memory needed for UnsafeInMemorySorter should be less than longArray in map.
-      map.freeArray();
-      // The memory used by UnsafeInMemorySorter will be counted later (end of this block)
       final UnsafeInMemorySorter inMemSorter = new UnsafeInMemorySorter(
-        taskMemoryManager, recordComparator, prefixComparator, Math.max(1, map.numElements()));
+        taskMemoryManager, recordComparator, prefixComparator, map.getArray());
```
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: Is there any chance that multiple objects may change the same array?

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java [hunk 15]
```diff
@@ -122,11 +119,7 @@ public UnsafeKVExternalSorter(
         /* initialSize */ 4096,
         pageSizeBytes,
         inMemSorter);
-
-      sorter.spill();
-      map.free();
-      // counting the memory used UnsafeInMemorySorter
-      taskMemoryManager.acquireExecutionMemory(inMemSorter.getMemoryUsage(), sorter);
+      map.reset();
```
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: I think we need comments at here. Basically, we need to explain why we reset. Also, once `sorter` is created, `inMemSorter` will not be needed, right?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java (hunk 12 above)
Comment (yhuai) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: ping @JoshRosen

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java [hunk 16]
```diff
@@ -243,9 +243,8 @@ public void printPerfMetrics() {
    * Note that this destroys the map, and as a result, the map cannot be used anymore after this.
    */
   public UnsafeKVExternalSorter destructAndCreateExternalSorter() throws IOException {
-    UnsafeKVExternalSorter sorter = new UnsafeKVExternalSorter(
+    return new UnsafeKVExternalSorter(
       groupingKeySchema, aggregationBufferSchema,
       SparkEnv.get().blockManager(), map.getPageSizeBytes(), map);
```
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java]: Is this map usable after this call?

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java (hunk 13 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: No, we check the space in page later.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java (hunk 16 above)
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java]: Also, I think we need to make it clear that `map` will not be freed at here.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java (hunk 14 above)
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: No, the array from map is not thread safe, only accessible from current thread.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java (hunk 15 above)
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: Yes.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java (hunk 14 above)
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: ok. Then, this `UnsafeInMemorySorter` is the only object that can hold the reference other than the original bytes to bytes map, right?

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java (hunk 12 above)
Comment (davies) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: Chatted with @JoshRosen offline, we should not clear `spillWriters` here.

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java (hunk 14 above)
Comment (davies) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeKVExternalSorter.java]: yes

File: core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java (hunk 12 above)
Comment (yhuai) [inline core/src/main/java/org/apache/spark/util/collection/unsafe/sort/UnsafeExternalSorter.java]: just a note, we had a quick discussion. Seems we should not call `spillWriters.clear()`. Otherwise those spilled files will not be deleted.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala (hunk 10 above)
Comment (davies) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMapSuite.scala]: The old behavior will be used anymore, so all the tests are updated (to reflect the real use case).

**Comment by yhuai** [comment]
test this please

File: sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java (hunk 16 above)
Comment (yhuai) [inline sql/core/src/main/java/org/apache/spark/sql/execution/UnsafeFixedWidthAggregationMap.java]: ok. I see, we cannot directly insert records to the external sorter returned by `destructAndCreateExternalSorter`. We need to use merge.

**Comment by yhuai** [comment]
LGTM pending jenkins.

**Comment by davies** [comment]
Merging into master, thanks!
